### ISSUE TITLE
MP's player list does not match the local player's state (in vehicle or not)

### ISSUE THREAD

**Issue by WillyB1337** [RigsOfRods/rigs-of-rods#2619]
### **Follow-up of November 2nd's discussion in #developpment channel**
_Hope I'm going to be clearer._

The color-shift is a new feature as It is told. The UI title bar and player name change color depending on the player's state.

> Notice the title bar and player name color

![Text color changes when in vehicle](https://user-images.githubusercontent.com/16364065/97921872-92482b80-1d29-11eb-8841-6a1dcd7f526e.gif)

Basically, the color shift feature when I get in or out of vehicles doesn't update the player's net label's color.

> Look at UnnamedPlayer's UI name vs netlabel name

![Head MP name dont change](https://user-images.githubusercontent.com/16364065/97921391-e1da2780-1d28-11eb-9154-2fd89a7a66c9.gif)

Also, it doesn't affect every color. As of my own testing, I determined that these colors are affected :
- Dark Yellow
- Bright Yellow
- Orange
- Bright Red
- Yellow
- Light Blue
- Light Orange

![image](https://user-images.githubusercontent.com/16364065/97922786-03d4a980-1d2b-11eb-812f-57c405def850.png)

The fact that only some colors are affected is a bit annoying to keep track of players.

The color doesn't update the netlabel player name depending on the local player's state. (keeps the player not in a vehicle state)
![image](https://user-images.githubusercontent.com/16364065/97923274-bb69bb80-1d2b-11eb-8f7b-706fd51ed5d1.png)

**Comment by tritonas00** [comment]
@only-a-ptr maybe i broke it?
https://github.com/RigsOfRods/rigs-of-rods/pull/2399/commits/12efddb313c4e286d120e5ae7dac6cc2e13287b9#diff-1070187583901765339bdd6f73ea22fd2afdf6e63222862f34527f7b4d9ada51

**Comment by only-a-ptr** [comment]
Sorry about the delayed response. I tried reproducing it but failed - the player colors didn't move for me even once. I tried entering/exiting vehicles and (dis)connecting repeatedly, no luck. Maybe it depends on terrain/vehicle.
![image](https://user-images.githubusercontent.com/491088/101251282-1cffba00-371e-11eb-8394-01a1d71b23de.png)

The alarming thing here is the color-changing window titlebar in first post. RoR has no such color-changing feature, the color isn't even configurable, it's hardcoded. If it changes anyway, it smells like memory corruption.

**Comment by only-a-ptr** [comment]
Re-opening to continue researching the root cause. @Zentro's patch may stop it from happening (if you could reproduce it in the first place, I couldn't) but it doesn't technically make sense.

## How it works

There's a hardcoded table of colors: https://github.com/RigsOfRods/ror-server/blob/master/source/server/sequencer.cpp#L224-L242 - total 25 colors (or 29 after Zentro's patch).

Each MP player has 'colour number' and RoR transforms it to actual color using this function: https://github.com/RigsOfRods/rigs-of-rods/blob/dd20e39ae2f36d009f43618b56ff3dd1aa272dcb/source/main/network/Network.cpp#L94-L100 - if it's a valid color number, the color is used, otherwise pitch black is used. This is safe.

Colour number is assigned by rorserver which looks for first unused color number, starting from 0 (valid): https://github.com/RigsOfRods/ror-server/blob/master/source/server/sequencer.cpp#L224-L242 - it doesn't have any upper limit, but as explained above, invalid colournum will just result in black color.

**Comment by WillyB1337** [comment]
So It looks like to be for that whole time the Reshade module does something to IMGUI to cause the color switch.
I was using the latest version of Reshade at the time, which is still 4.8.2
I do have Reshade running here.
![screenshot_2020-12-05_14-51-13_1](https://user-images.githubusercontent.com/16364065/101262334-df923100-370b-11eb-9609-a6e120090a07.png)
When in a vehicle, the color switch occurs.
![screenshot_2020-12-05_14-51-31_1](https://user-images.githubusercontent.com/16364065/101262339-ee78e380-370b-11eb-8c4d-248a686d0219.png)
Even when the presets are disabled, the effect is still happening.
![screenshot_2020-12-05_14-52-10_1](https://user-images.githubusercontent.com/16364065/101262347-081a2b00-370c-11eb-937a-75076d884df6.png)

I did tests without Reshade and clean RoR.cfg and the mentioned behavior never happened.

**On the pictures, I have an FPS counter and system clock which are not present in RoR normally, I would have hoped anyone would have noticed that.**

I do now realize that I should have indicated in my testing that Reshade was active.
_Back then, I did the pictures thru RoR's SAVE_SCREENSHOT, which don't show the presets but still displays the Reshade's hud._

I would like to give my apologies for all the confusion I gave to anyone that tried to look into the code for no reason. I believe that since the behavior happens from an external source which we cannot control, we would look into either the IMGUI code we're using or just close this issue and considered it an outside source bug.
